### ISSUE TITLE
piet.clear rework

### ISSUE THREAD

**Issue by Ciantic** [linebender/piet#399]
We noticed this with @JAicewizard. We are discussing [this in Zulip](https://xi.zulipchat.com/#narrow/stream/147926-druid/topic/semi-transparent.20.2F.20piet.2Eclear.20rework)

Piet's `clear` method is usually useless because if your background is opaque, it's not useful as the background color of the window overwrites this, the bug appears only if your background has alpha channel (transparent, semi-transparent).

On d2d side the `piet.clear` calls this: [ID2D1RenderTarget::Clear](
https://docs.microsoft.com/en-us/windows/win32/api/d2d1/nf-d2d1-id2d1rendertarget-clear(constd2d1_color_f_)), the docs says:

> If the render target has an active clip (specified by PushAxisAlignedClip), the clear command is applied only to the area within the clip region.

But Piet doesn't use this way of setting the clipping at the moment.

However, we figure that fixing `clear` on Windows requires similar implementantion as e.g. in Core graphics side:

```rust
    fn clear(&mut self, color: Color) {
        let (r, g, b, a) = color.as_rgba();
        self.ctx.set_rgb_fill_color(r, g, b, a);
        self.ctx.set_blend_mode(CGBlendMode::Copy);
        self.ctx.fill_rect(self.ctx.clip_bounding_box());
        self.ctx.set_blend_mode(self.blend_mode);
    }
```
(See #398)

Windows clear must be made to use similar method with D2D.

Part of the motivation is to get rid of this hack I made: https://github.com/linebender/druid/blob/131b0a6d0622e0a6dad3c0a8945bec88df1f0f3f/druid-shell/src/platform/windows/window.rs#L400-L424

**Comment by JAicewizard** [comment]
(again, its coregraphics not cairo ;))

**Comment by raphlinus** [comment]
We should definitely set up axis aligned clipping when the clip shape is in fact an axis aligned rectangle - that should be a big performance improvement, and will increase the cases that Cleartype rendering is enabled.

It gets us to a funny place though, where `clear` has different semantics depending on whether the clip shape is an axis aligned rectangle or no. I think it's possible to document that, as it fits with the primary use case.

**Comment by Ciantic** [comment]
Currently as things stand, is that:

Idea is that clear overwrites the alpha channel with given color on the given rect.

@cmyr created a starting point:

https://github.com/linebender/piet/pull/401

The `clear` gets a `rect` input that is used to blend to COPY, or do what ever to make the color match the given color including alpha, within that rect.

This can then be used to invalidate rects even to color of are transparent.

**Comment by Ciantic** [comment]
I'll try to sum this up, we think with @JAicewizard that this is correct render:

![image](https://user-images.githubusercontent.com/64731/108638183-9a71d580-7496-11eb-9ee3-775a2ee7d16a.png)

Just like in @cmyr's example. However we don't want to ignore translate, so let's change `picture_15.rs` comments from this:

```rust

    // this should ignore the translation and the clipping, and clear everything
    rc.clear(None, SEMI_TRANSPARENT_WHITE);

    // this should ignore the translate and clipping and clear just the rect
    rc.clear(rect, BLUE);
```

it should say:

```rust
    // this should ignore the clipping, and clear everything
    rc.clear(None, SEMI_TRANSPARENT_WHITE);

    // this should ignore the clipping and clear just the rect
    rc.clear(rect, BLUE);
```

As we don't want to ignore translate or any other things about transform.

**Comment by cmyr** [comment]
So I don't think I quite agree.

For instance: I don't think two windows sharing a canvas is a realistic concern.

I also think that if DPI is included in the reported transform, that's a problem on the piet side. How I would think about this:

- clear should ignore transforms and clipping *that have been applied by piet*. If the underlying platform render object has transforms when it is incoming, we should never touch these.

I'm actually not sure to what extent this matters, but it should be consistent on the different platforms. `current_transform()` should always be `Affine::IDENTITY` for a newly created render context.

**Comment by cmyr** [comment]
closed with #401
